src/main/scala/org/constellation/primitives/concurrency/SingleRef.scala

```diff
@@ -0,0 +1,34 @@
+package org.constellation.primitives.concurrency
+
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.{Concurrent, IO}
+
+class SingleRef[F[_]: Concurrent, A](default: A) {
+
+  private val ref = Ref.unsafe[F, A](default)
+  private val semaphore: Semaphore[F] = Semaphore.in[IO, F](1).unsafeRunSync()
+
+  val lock = new SingleLock[F, A]("", semaphore)
+
+  def acquire: F[Unit] = lock.acquire
+  def release: F[Unit] = lock.release
+
+  def get: F[A] = withLock(ref.get)
+  def getUnsafe: F[A] = ref.get
+
+  def set(v: A): F[Unit] = withLock(ref.set(v))
+  def setUnsafe(v: A): F[Unit] = ref.set(v)
+
+  def update(f: A => A): F[Unit] = withLock(ref.update(f))
+  def updateUnsafe(f: A => A): F[Unit] = (ref.update(f))
+
+  def modify[B](f: A => (A, B)): F[B] = withLock(ref.modify(f))
```

Comment:
If `modify` is atomic why we need to wrap it with lock ?

Reply:
It's not about being atomic or not but to have the exclusive access, which modify don't support.